allennlp_server/commands/configure.py

```diff
@@ -0,0 +1,54 @@
+"""
+The ``configure`` subcommand launches a webapp that helps you generate an AllenNLP configuration file.
+
+    $ allennlp configure --help
```

Comment:
Is this usage still correct? I'm not sure how allennlp will know to include this subcommand without something changing in the main repo? Maybe i'm missing the point...

Reply:
Yeah, there are more details in the PR description. But the main idea is that the allennlp  main codebase checks for the presence of allennlp-server by importing it, and adds the subcommands. Does it make sense? Other ideas? Another idea that comes to my mind is a separate command.

Reply:
Yes, I agree with this.  One option for doing this is to have something like a `.allennlp/plugins` file, where we import everything in that file before running our command.  This lets you add whatever package you want and removes the need for `--include-package`.  The pip install script could add a line to the plugin file.

Reply:
I searched a bit on this and found out that this is a thing: [Creating and discovering plugins](https://packaging.python.org/guides/creating-and-discovering-plugins/). The most convincing method to me is the 3rd one, [Using package metadata](https://packaging.python.org/guides/creating-and-discovering-plugins/#using-package-metadata), as the first one limits to packages starting with "allennlp" (maybe it's desired?) and the second one is complicated as it requires following certain directory hierarchy pattern and then adding packages.

So it'd be about declaring an entry point in the setup.py and then making allennlp import it, adding any available subcommand. Does it make sense? I'm gonna start working on it, and we can change it later if we decide.

Reply:
I've analyzed it more. The problem I see with `.allennlp/plugins` is that any Python installation and virtual env in the system would read the same file, and for some of them, the plugins may not be installed.

For the solution I'm proposing, to remove the need for `--include-package`, the AllenNLP user has to include its code as a plugin, which involves installing its own package, which is cumbersome.

I think that using a standard namespace for it would be ideal. So, people can put some imports under a file in a certain folder to have their files discovered automatically. Python doesn't seem to support this without some packaging setup. We can import whatever is inside a certain folder, [like this old SO answer](https://stackoverflow.com/a/932072/1165181). I also found a library called [pluginlib](https://github.com/Rockhopper-Technologies/pluginlib), which seems to be somewhat used and maintained, which facilitates this. `Subcommand` can be the plugin interface, and then any subclasses are loaded. To also implement the `--include-package` behavior, other classes such as `Model` could also be plugins.

We can try just doing the first thing (defining a standard folder such as `allennlp.plugins` and loading whatever is there) and see how it goes.

Thoughts?

Reply:
I think all we need to do is try to import every module in the plugins file. It's one package name per line, as a string. We just call exactly the same code that the current `--include-package` logic calls. If the import fails, due to a path error or a virtualenv issue as you brought up, we just give a big warning and move on. It's the job of the person who designed the package to make sure that everything that should be registered gets imported. And if our `Subcommands` aren't currently loaded from the registry, they should be.

This simple solution solves all of the issues, doesn't it?

Reply:
I guess there's the question of how a line gets added to the plugins file on pip install, but I assume that's solvable. If a user wants to avoid using `--include-package` all the time, they can manually add a line to the plugins file.

Reply:
> It's the job of the person who designed the package to make sure that everything that should be registered gets imported. And if our Subcommands aren't currently loaded from the registry, they should be.

Agree.

> This simple solution solves all of the issues, doesn't it?

Yes, but one new issue I see: getting warnings in one virtualenv that uses allennlp for a plugin installed within another one. We could avoid showing warnings but silent errors are worse than noise when you want the feature to work.

I think that instead of trying to import from this global place, we could try to import all that is under `allennlp.plugins`, whatever is there. The core idea is the same as you say but with things local to the virtualenv instead of global to the system.

Reply:
How does something get added to `allennlp.plugins`?  How does a user see what plugins are there, and add one manually?

Reply:
I'm gonna try and see if it works, but to add: add something to the folder (python package) `allennlp/plugins/`. To see what plugins are there: look for submodules of `allennlp.plugins`.

Reply:
I would strongly prefer the other approach. I don't know where I would even go to find that folder in my environment, I'd have to dig around, so we can't give consistent instructions to users.  And those are automatically modified by pip in ways that I couldn't tell you, so I don't know if we would get clobbered on an update, or something.  *Much* better to just have a plain text file in a consistent location.

Reply:
Okay.

Just as a comment, I tested what I said but [it doesn't work with regular packages, only with namespace packages](https://docs.python.org/3/reference/import.html#packages), which is [one way Python recommends for plugin discoverability](https://packaging.python.org/guides/creating-and-discovering-plugins/#using-namespace-packages). But it seems users would need to [have a setup.py and use the attribute `packages`](https://packaging.python.org/guides/packaging-namespace-packages/#native-namespace-packages).